Win compilation fails due to some missing includes and a type size (fixed)

Hi,

I tried to compile Latruculi under Win11 with msys2 and got those errors below:

```
....
[ 34%] Building CXX object CMakeFiles/benchmark.dir/bench/benchmark.cpp.obj
C:/msys64/home/Admin/latrunculi/bench/benchmark.cpp:29:17: error: use of undeclared identifier 'MAX_PATH'
   29 |     char buffer[MAX_PATH];
      |                 ^~~~~~~~
C:/msys64/home/Admin/latrunculi/bench/benchmark.cpp:30:37: error: use of undeclared identifier 'MAX_PATH'
   30 |     GetModuleFileName(NULL, buffer, MAX_PATH);
      |                                     ^~~~~~~~
C:/msys64/home/Admin/latrunculi/bench/benchmark.cpp:31:5: error: use of undeclared identifier 'exePath'; did you mean 'exe_path'?
   31 |     exePath = buffer;
      |     ^~~~~~~
      |     exe_path
C:/msys64/home/Admin/latrunculi/bench/benchmark.cpp:26:27: note: 'exe_path' declared here
   26 |     std::filesystem::path exe_path;
      |                           ^
[ 36%] Linking CXX executable bin/latrunculi.exe
3 errors generated.
...
```

Fixed those by including windows.h and renaming exePath to exe_path.

Then I got another error about the size of long not sufficient for some numbers in a type and changed it to long long.

`using PerftParams = std::tuple<std::string, std::vector<long long>>;`

```
...
[ 95%] Built target benchmark
C:/msys64/home/Admin/latrunculi/tests/movegen.test.cpp:35:9: error: constant expression evaluates to 8031647685 which cannot be narrowed to type 'value_type' (aka 'long') [-Wc++11-narrowing]
   35 |         8031647685,
      |         ^~~~~~~~~~
C:/msys64/home/Admin/latrunculi/tests/movegen.test.cpp:35:9: note: insert an explicit cast to silence this issue
   35 |         8031647685,
      | [ 97%] Linking CXX static library ../../../lib/libgmock_main.a
        ^~~~~~~~~~
      |         static_cast<value_type>( )
C:/msys64/home/Admin/latrunculi/tests/movegen.test.cpp:35:9: warning: implicit conversion from 'long long' to 'const value_type' (aka 'const long') changes value from 8031647685 to -558286907
      [-Wconstant-conversion]
   29 |     {
      |     ~
   30 |         48,
   31 |         2039,
   32 |         97862,
   33 |         4085603,
   34 |         193690690,
   35 |         8031647685,
      |         ^~~~~~~~~~
C:/msys64/home/Admin/latrunculi/tests/movegen.test.cpp:94:9: error: constant expression evaluates to 6923051137 which cannot be narrowed to type 'value_type' (aka 'long') [-Wc++11-narrowing]
   94 |         6923051137,
      |         ^~~~~~~~~~
C:/msys64/home/Admin/latrunculi/tests/movegen.test.cpp:94:9: note: insert an explicit cast to silence this issue
   94 |         6923051137,
      |         ^~~~~~~~~~
      |         static_cast<value_type>( )
C:/msys64/home/Admin/latrunculi/tests/movegen.test.cpp:94:9: warning: implicit conversion from 'long long' to 'const value_type' (aka 'const long') changes value from 6923051137 to -1666883455
      [-Wconstant-conversion]
...
```

build with -O3 -march=native -flto
```
uci

id name Latrunculi 0.1.0
id author Eric VanderHelm
option name Hash type spin default 4 min 1 max 2048
option name Threads type spin default 1 min 1 max 64
uciok
isready
readyok
position startpos
go depth 15
info depth 1 score cp 86 nodes 33 time 0 nps 0 pv e2e4
info depth 2 score cp 20 nodes 159 time 0 nps 0 pv e2e4 e7e5
info depth 3 score cp 63 nodes 786 time 0 nps 0 pv e2e4 e7e5 b1c3
info depth 4 score cp 20 nodes 1696 time 1 nps 1696000 pv e2e4 e7e5 b1c3 b8c6
info depth 5 score cp 28 nodes 4790 time 2 nps 2395000 pv e2e4 e7e5 b1c3 b8c6 d2d3
info depth 6 score cp 30 nodes 10993 time 4 nps 2748250 pv b1c3 b8c6 d2d4 e7e6 e2e4 f8b4
info depth 7 score cp 34 nodes 25957 time 10 nps 2595700 pv b1c3 b8c6 d2d4 g8f6 g1f3 d7d5 e2e3
info depth 8 score cp 18 nodes 60871 time 24 nps 2536291 pv d2d4 b8c6 d4d5 c6e5 g1f3 d7d6 f3e5 d6e5
info depth 9 score cp 41 nodes 118964 time 47 nps 2531148 pv d2d4 d7d5 b1c3 b8c6 g1f3 e7e6 e2e3 f8b4 f1b5
info depth 10 score cp 20 nodes 207629 time 82 nps 2532060 pv d2d4 d7d5 b1c3 b8c6 g1f3 e7e6 e2e3 g8f6 f1b5 f8b4
info depth 11 score cp 33 nodes 353922 time 140 nps 2528014 pv d2d4 d7d5 b1c3 b8c6 g1f3 e7e6 e2e4 g8f6 e4d5 f6d5 c1g5
info depth 12 score cp 22 nodes 799381 time 324 nps 2467225 pv d2d4 d7d5 b1c3 e7e6 g1f3 g8f6 c1f4 f8b4 a2a3 b4c3 b2c3 e8g8
info depth 13 score cp 30 nodes 1503071 time 622 nps 2416512 pv b1c3 d7d5 d2d4 e7e6 e2e3 g8f6 g1f3 f8e7 f1d3 e8g8 e1g1 b8c6 f3g5
info depth 14 score cp 34 nodes 2328744 time 973 nps 2393364 pv b1c3 d7d5 d2d4 e7e6 e2e3 g8f6 g1f3 b8c6 f1b5 f8d6 f3e5 e8g8 e5c6
info depth 15 score cp 32 nodes 4768548 time 2018 nps 2363006 pv g1f3 g8f6 e2e3 d7d5 b1c3 b8c6 d2d4 e7e6 f1e2 f8e7 e1g1 e8g8 c3d5 f6d5
bestmove g1f3
```

BTW what is the cmake command for just building latrunculi? (w/o the test and bench binary)

Guenther (rwbc)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -5,6 +5,7 @@ set(CMAKE_CXX_STANDARD 23)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_EXTENSIONS OFF)
 
+option(BUILD_TESTING "Build tests" OFF)
 include(CTest)
 find_package(Threads REQUIRED)
 
@@ -35,7 +36,7 @@ target_compile_definitions(latrunculi_lib PRIVATE LATRUNCULI_VERSION="${PROJECT_
 add_executable(latrunculi src/main.cpp)
 target_link_libraries(latrunculi PRIVATE latrunculi_lib Threads::Threads)
 
-option(LATRUNCULI_BUILD_BENCHMARK "Build benchmark target" ON)
+option(LATRUNCULI_BUILD_BENCHMARK "Build benchmark target" OFF)
 if(LATRUNCULI_BUILD_BENCHMARK)
     add_executable(benchmark bench/benchmark.cpp)
     target_link_libraries(benchmark PRIVATE latrunculi_lib Threads::Threads)

--- a/CMakePresets.json
+++ b/CMakePresets.json
@@ -25,12 +25,32 @@
       }
     },
     {
+      "name": "debug-dev",
+      "inherits": ["debug"],
+      "displayName": "Debug Developer",
+      "description": "Debug build with tests and benchmark enabled",
+      "cacheVariables": {
+        "BUILD_TESTING": "ON",
+        "LATRUNCULI_BUILD_BENCHMARK": "ON"
+      }
+    },
+    {
       "name": "release",
       "inherits": ["base"],
       "displayName": "Release",
       "description": "Release build using the system default C++ compiler",
       "cacheVariables": {
         "CMAKE_BUILD_TYPE": "Release"
+      }
+    },
+    {
+      "name": "release-dev",
+      "inherits": ["release"],
+      "displayName": "Release Developer",
+      "description": "Release build with tests and benchmark enabled",
+      "cacheVariables": {
+        "BUILD_TESTING": "ON",
+        "LATRUNCULI_BUILD_BENCHMARK": "ON"
       }
     },
     {
@@ -41,6 +61,16 @@
       "cacheVariables": {
         "CMAKE_BUILD_TYPE": "Release",
         "LATRUNCULI_SEARCH_STATS": "ON"
+      }
+    },
+    {
+      "name": "release-stats-dev",
+      "inherits": ["release-stats"],
+      "displayName": "Release with search stats Developer",
+      "description": "Release build with search stats, tests, and benchmark enabled",
+      "cacheVariables": {
+        "BUILD_TESTING": "ON",
+        "LATRUNCULI_BUILD_BENCHMARK": "ON"
       }
     }
   ],
@@ -59,26 +89,41 @@
       "name": "release-stats",
       "configurePreset": "release-stats",
       "jobs": 0
+    },
+    {
+      "name": "debug-dev",
+      "configurePreset": "debug-dev",
+      "jobs": 0
+    },
+    {
+      "name": "release-dev",
+      "configurePreset": "release-dev",
+      "jobs": 0
+    },
+    {
+      "name": "release-stats-dev",
+      "configurePreset": "release-stats-dev",
+      "jobs": 0
     }
   ],
   "testPresets": [
     {
-      "name": "debug",
-      "configurePreset": "debug",
+      "name": "debug-dev",
+      "configurePreset": "debug-dev",
       "output": {
         "outputOnFailure": true
       }
     },
     {
-      "name": "release",
-      "configurePreset": "release",
+      "name": "release-dev",
+      "configurePreset": "release-dev",
       "output": {
         "outputOnFailure": true
       }
     },
     {
-      "name": "release-stats",
-      "configurePreset": "release-stats",
+      "name": "release-stats-dev",
+      "configurePreset": "release-stats-dev",
       "output": {
         "outputOnFailure": true
       }

--- a/README.md
+++ b/README.md
@@ -28,11 +28,28 @@ Tested on linux with g++ 13.3 / 14.2 and clang++ 18.
 
 ### build
 
+Default builds produce the engine executable only:
+
 ```bash
 git clone --recursive https://github.com/zm-bm/latrunculi.git
 cd latrunculi
 cmake --preset release
 cmake --build --preset release
+```
+
+Developer builds opt into tests and benchmark:
+
+```bash
+cmake --preset release-dev
+cmake --build --preset release-dev
+ctest --preset release-dev
+```
+
+Specific development targets can be built directly:
+
+```bash
+cmake --build --preset release-dev --target tests
+cmake --build --preset release-dev --target benchmark
 ```
 
 if you want a specific compiler:
@@ -77,10 +94,12 @@ go depth 20
 ## tests
 
 ```bash
-ctest --preset release
+cmake --preset release-dev
+cmake --build --preset release-dev
+ctest --preset release-dev
 ```
 
-`benchmark` is available for quick smoke checks during development.
+`benchmark` is available in `*-dev` presets for quick smoke checks.
 
 ## license
 

--- a/bench/benchmark.cpp
+++ b/bench/benchmark.cpp
@@ -8,7 +8,12 @@
 #include <sstream>
 #include <string>
 #include <thread>
+
+#ifndef _WIN32
+#include <limits.h>
 #include <unistd.h>
+#endif
+
 #include <vector>
 
 #include "bench_defs.hpp"
@@ -20,14 +25,15 @@ std::string TESTFILE;
 int         SEARCHTIME = 1000; // smoke benchmark search time in milliseconds
 const int   HASH_MB    = 16;
 int         THREADS    = 1;
-int         MOVE_LIMIT = 5;  // max moves to search in a smoke run
+int         MOVE_LIMIT = 5; // max moves to search in a smoke run
 
 // Get test file path relative to the executable
 static std::string get_test_file_path() {
     std::vector<std::filesystem::path> candidates;
 
 #ifdef LATRUNCULI_SOURCE_DIR
-    candidates.emplace_back(std::filesystem::path(LATRUNCULI_SOURCE_DIR) / "bench" / "arasan20.epd");
+    candidates.emplace_back(std::filesystem::path(LATRUNCULI_SOURCE_DIR) / "bench" /
+                            "arasan20.epd");
 #endif
 
 #ifndef _WIN32
@@ -141,7 +147,8 @@ public:
                 std::string        token;
                 bestmove_line >> token >> result.bestmove;
                 result.has_bestmove = !result.bestmove.empty();
-                result.has_plausible_bestmove = result.is_plausible_move(testCase.fen, result.bestmove);
+                result.has_plausible_bestmove =
+                    result.is_plausible_move(testCase.fen, result.bestmove);
                 continue;
             }
 
@@ -180,33 +187,35 @@ public:
             observed_nodes.push_back(result.observed_max_nodes);
             observed_nps.push_back(result.observed_nps);
             completed_cases += result.saw_observed_info ? 1 : 0;
-            bestmove_cases += result.has_bestmove ? 1 : 0;
+            bestmove_cases  += result.has_bestmove ? 1 : 0;
             plausible_cases += result.has_plausible_bestmove ? 1 : 0;
         }
 
         const auto [min_depth, max_depth] = minmax_of(observed_depths);
-        const auto [min_time, max_time] = minmax_of(observed_times);
+        const auto [min_time, max_time]   = minmax_of(observed_times);
         const auto [min_nodes, max_nodes] = minmax_of(observed_nodes);
-        const auto [min_nps, max_nps] = minmax_of(observed_nps);
+        const auto [min_nps, max_nps]     = minmax_of(observed_nps);
 
         std::cout << "\nSmoke Benchmark Summary\n";
         std::cout << "Threads = " << THREADS << ", ";
         std::cout << "movetime = " << SEARCHTIME << " ms, ";
         std::cout << "cases = " << observed_depths.size() << std::endl;
         std::cout << "-------------------" << std::endl;
-        std::cout << "Completed with observed info: " << completed_cases << " / " << observed_depths.size()
-                  << ", bestmove seen: " << bestmove_cases << " / " << observed_depths.size()
-                  << ", plausible bestmove: " << plausible_cases << " / " << observed_depths.size() << std::endl;
-        std::cout << "Observed averages: depth " << format_metric(average_of(observed_depths)) << " ply, nodes "
-                  << format_metric(average_of(observed_nodes)) << ", time "
+        std::cout << "Completed with observed info: " << completed_cases << " / "
+                  << observed_depths.size() << ", bestmove seen: " << bestmove_cases << " / "
+                  << observed_depths.size() << ", plausible bestmove: " << plausible_cases << " / "
+                  << observed_depths.size() << std::endl;
+        std::cout << "Observed averages: depth " << format_metric(average_of(observed_depths))
+                  << " ply, nodes " << format_metric(average_of(observed_nodes)) << ", time "
                   << format_metric(average_of(observed_times)) << " ms, nps "
                   << format_metric(average_of(observed_nps)) << std::endl;
-        std::cout << "Observed medians: nps " << format_metric(median_of(observed_nps)) << ", depth "
-                  << format_metric(median_of(observed_depths)) << " ply" << std::endl;
+        std::cout << "Observed medians: nps " << format_metric(median_of(observed_nps))
+                  << ", depth " << format_metric(median_of(observed_depths)) << " ply" << std::endl;
         std::cout << "Observed spread: depth " << min_depth << "-" << max_depth << " ply, nodes "
                   << min_nodes << "-" << max_nodes << ", time " << min_time << "-" << max_time
                   << " ms, nps " << min_nps << "-" << max_nps << std::endl;
-        std::cout << "Use this as a quick stability/throughput smoke benchmark, not as a tactical or comparative suite."
+        std::cout << "Use this as a quick stability/throughput smoke benchmark, not as a tactical "
+                     "or comparative suite."
                   << std::endl;
     }
 };
@@ -263,7 +272,8 @@ int main(int argc, char* argv[]) {
 
         if (!TESTFILE.empty()) {
             if (!std::filesystem::exists(TESTFILE)) {
-                std::cerr << "Warning: Provided test file does not exist: " << TESTFILE << std::endl;
+                std::cerr << "Warning: Provided test file does not exist: " << TESTFILE
+                          << std::endl;
                 std::cerr << "Falling back to automatic path resolution." << std::endl;
                 TESTFILE = get_test_file_path();
             }

--- a/tests/movegen.test.cpp
+++ b/tests/movegen.test.cpp
@@ -1,5 +1,7 @@
 #include "movegen.hpp"
 
+#include <cstdint>
+
 #include <gtest/gtest.h>
 
 #include "board.hpp"
@@ -10,7 +12,7 @@
 
 constexpr auto depth_limit = 4;
 
-using PerftParams = std::tuple<std::string, std::vector<long>>;
+using PerftParams = std::tuple<std::string, std::vector<uint64_t>>;
 
 PerftParams perft_start = {
     STARTFEN,
@@ -108,7 +110,7 @@ TEST_P(PerftTest, PerftDepths) {
         if (depth > depth_limit)
             break;
 
-        long result = board.perft(depth, null_stream);
+        uint64_t result = board.perft(depth, null_stream);
         EXPECT_EQ(result, expected[depth - 1]) << "Failed at depth " << depth;
     }
 }

